src/staking.cairo

```diff
@@ -1,325 +1,452 @@
 
-            assert(!res.withdrawn, 'stake withdrawn already');
+            assert(old_locked.amount > 0, 'No existing lock found');
+            assert(old_locked.end > current_time, 'Lock expired');
+            assert(unlock_date > old_locked.end, 'Can only increase lock duration');
+            assert(unlock_date <= current_time + MAXTIME, 'Voting lock can be 4 years max');
 
-            assert(res.amount_staked != 0, 'no stake found, check stake id');
-            let unlock_date = res.start_date + res.length;
-            assert(get_block_timestamp() > unlock_date, 'unlock time not yet reached');
+            let old_slope = old_locked.amount / (old_locked.end - current_time).into();
+            let new_slope = old_locked.amount / (unlock_date - current_time).into();
 
-            let voting_token = IERC20Dispatcher {
-                contract_address: get_governance_token_address_self()
-            };
-            voting_token.burn(caller, res.amount_voting_token.into());
+            self.total_slope.write(self.total_slope.read() - old_slope + new_slope);
+
+            let new_locked = LockedBalance { amount: old_locked.amount, end: unlock_date };
+
+            self.locked.write(caller, new_locked);
+            self._checkpoint(caller, old_locked, new_locked);
 
-            let floating_token = IERC20Dispatcher {
-                contract_address: self.floating_token_address.read()
-            };
-            // user gets back the same amount of tokens they put in.
-            // the payoff is in holding voting tokens, which make the user eligible for distributions of protocol revenue
-            // works for tokens with fixed max float
-            floating_token.transfer(caller, res.amount_staked.into());
-            self
-                .stake
-                .write(
-                    (caller, id),
-                    Stake {
-                        amount_staked: res.amount_staked,
-                        amount_voting_token: res.amount_voting_token,
-                        start_date: res.start_date,
-                        length: res.length,
-                        withdrawn: true
-                    }
-                );
             self
                 .emit(
-                    Unstaked {
-                        user: caller,
-                        stake_id: id,
-                        amount: res.amount_staked,
-                        amount_voting_token: res.amount_voting_token,
-                        start_date: res.start_date,
-                        length: res.length
+                    Deposit {
+                        caller,
+                        amount: 0,
+                        locktime: unlock_date,
+                        type_: DEPOSIT_TYPE_INCREASE_TIME,
+                        ts: current_time,
                     }
                 );
         }
+        fn withdraw(ref self: ComponentState<TContractState>, caller: ContractAddress) {
```

Comment:
I don't think you burn the locked token anywhere?

Reply:
if caller is withdrawing their staked tokens, shouldn't it be transferred back to them and not burned?